protovalidate/internal/extra_func.py

```diff
@@ -234,6 +221,737 @@ def unique(val: celtypes.Value) -> celpy.Result:
+                # > URI producing applications must not use percent-encoding in host
+                # > unless it is used to represent a UTF-8 character sequence.
+                if not self.__check_host_pct_encoded(raw_host):
+                    return False
+
+            return True
+
+        return False
+
+    def __port(self) -> bool:
+        """Determines whether string contains a valid port.
+
+        Method parses the rule:
+
+        port = *DIGIT
+
+        Terminated by end of authority.
+        """
+
+        start = self._index
+        while True:
+            if self.__digit():
+                continue
+
+            if self.__is_authority_end():
+                return True
+
+            self._index = start
+            return False
+
+    def __ip_literal(self) -> bool:
+        """Determines whether string contains a valid port.
+
+        Method parses the rule from RFC 6874:
+
+        IP-literal = "[" ( IPv6address / IPv6addrz / IPvFuture  ) "]"
+        """
+
+        start = self._index
+
+        if self.__take("["):
+            curr_idx = self._index
+            if self.__ipv6_address() and self.__take("]"):
+                return True
+
+            self._index = curr_idx
+
+            if self.__ipv6_addrz() and self.__take("]"):
+                return True
+
+            self._index = curr_idx
+
+            if self.__ip_vfuture() and self.__take("]"):
+                return True
+
+        self._index = start
+        return False
+
+    def __ipv6_address(self) -> bool:
+        """Determines whether string contains a valid ipv6 address.
+
+        Method parses the rule "IPv6address".
+
+        Relies on the implementation of validate_ip.
+        """
+
+        start = self._index
+        while self.__hex_dig() or self.__take(":"):
+            pass
+
+        if validate_ip(self._string[start : self._index], 6):
+            return True
+
+        self._index = start
+        return False
+
+    def __ipv6_addrz(self) -> bool:
+        """Determines whether string contains a valid IPv6addrz.
+
+        RFC 6874:
+
+        IPv6addrz = IPv6address "%25" ZoneID
+        """
+
+        start = self._index
+        if self.__ipv6_address() and self.__take("%") and self.__take("2") and self.__take("5") and self.__zone_id():
+            return True
+
+        self._index = start
+
+        return False
+
+    def __zone_id(self) -> bool:
+        """Determines whether string contains a valid zone ID.
+
+        RFC 6874:
+
+        ZoneID = 1*( unreserved / pct-encoded )
+        """
+
+        start = self._index
+        while self.__unreserved() or self.__pct_encoded():
+            pass
+
+        if self._index - start > 0:
+            return True
+
+        self._index = start
+
+        return False
+
+    def __ip_vfuture(self) -> bool:
+        """Determines whether string contains a valid ipvFuture.
+
+        IPvFuture  = "v" 1*HEXDIG "." 1*( unreserved / sub-delims / ":" )
```

Comment:
Above, we have a short sentence above the quoted ABNF rule:

```suggestion
        
        Method parses the rule:
        
        IPvFuture  = "v" 1*HEXDIG "." 1*( unreserved / sub-delims / ":" )
```

I think that's a good idea. We should use it consistently - here and at all other places.